modules/javafx.base/src/test/java/test/javafx/beans/value/ObservableValueWhenTest.java

```diff
@@ -0,0 +1,225 @@
+
+    @Nested
+    class WhenNotObserved {
+
+        @Nested
+        class AndConditionStartsFalse {
+            BooleanProperty condition = new SimpleBooleanProperty(false);
+
+            @Test
+            void shouldNeverCallDownstreamMapFunction() {
+                StringProperty property = new SimpleStringProperty("a");
+                List<String> observedMappings = new ArrayList<>();
+
+                property.when(condition).map(observedMappings::add);
+
+                assertEquals(List.of(), observedMappings);
+
+                property.set("b");
+
+                assertEquals(List.of(), observedMappings);
+
+                condition.set(true);
+
+                assertEquals(List.of(), observedMappings);
+
+                property.set("c");
+
+                assertEquals(List.of(), observedMappings);
+
+                condition.set(false);
+
+                assertEquals(List.of(), observedMappings);
+
+                property.set("d");
+
+                assertEquals(List.of(), observedMappings);
+
+                condition.set(true);
+
+                assertEquals(List.of(), observedMappings);
+            }
+        }
+
+        @Nested
+        class AndConditionStartsTrue {
+            BooleanProperty condition = new SimpleBooleanProperty(true);
+
+            @Test
+            void shouldNeverCallDownstreamMapFunction() {
+                StringProperty property = new SimpleStringProperty("a");
+                List<String> observedMappings = new ArrayList<>();
+
+                property.when(condition).map(observedMappings::add);
+
+                assertEquals(List.of(), observedMappings);
+
+                property.set("b");
+
+                assertEquals(List.of(), observedMappings);
+
+                condition.set(false);
+
+                assertEquals(List.of(), observedMappings);
+
+                property.set("c");
+
+                assertEquals(List.of(), observedMappings);
+
+                condition.set(true);
+
+                assertEquals(List.of(), observedMappings);
+
+                property.set("d");
+
+                assertEquals(List.of(), observedMappings);
+
+                condition.set(false);
+
+                assertEquals(List.of(), observedMappings);
+            }
+        }
+    }
+
+    @Nested
+    class WhenObserved {
+        @Nested
+        class AndConditionStartsFalse {
+            BooleanProperty condition = new SimpleBooleanProperty(false);
+
+            @Test
+            void shouldCallDownstreamMapFunctionOnlyWhenAbsolutelyNecessary() {
+                StringProperty property = new SimpleStringProperty("a");
+                List<String> observedMappings = new ArrayList<>();
+                List<String> observedChanges = new ArrayList<>();
+
+                property.when(condition)
+                    .map(x -> { observedMappings.add(x); return x; })
+                    .addListener((obs, old, current) -> observedChanges.add(old + " -> " + current));
```

Comment:
What do you think about a variant with an invalidation listener?

Reply:
That's not really useful as there is no way to make such a fluent binding valid again (it would invalidate once, and never become valid).  The only way to make it valid would be to assign the result of `map` to a variable, and call `get` on that.

I can however add a test that proves the above :)

Reply:
Also, the test is specifically looking how often `map` is called here (which is triggered by invalidations from the `when`).  The listener is only there to make sure that `map` has installed its listeners as fluent observables don't install them when not observed themselves.  I could have used an `InvalidationListener` here for this purpose, but then I'd still need to call `getValue` on the `map` observable as otherwise it will definitely not map anything as it is invalid.

Perhaps a simple test that shows this is the case can be useful.

Reply:
I've added a variant with an InvalidationListener directly on the when binding.  It works as I would expect, it's maybe a more useful test than the others (it basically tests the same thing, but more directly).